File: 1022-sum-of-root-to-leaf-binary-numbers/1022-sum-of-root-to-leaf-binary-numbers.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int sumRootToLeaf_01(TreeNode node,int sum) {
        //Case when our node itself is null so tree doesnt exist.
        if(node == null) return 0;
        //calculate Pre sum.
        sum = (2*sum) + node.val;
        //Condition for recognizing leaf
        if(node.left ==null && node.right ==null) return sum;
        //left Whole left & right subtree's decimal sum
        int leftSum = sumRootToLeaf_01(node.left,sum);
        int rightSum = sumRootToLeaf_01(node.right,sum);
        //return this total Sum.
        int totalSum = leftSum + rightSum;
        
        return totalSum;
    }
    
    public int sumRootToLeaf(TreeNode root) {
        return sumRootToLeaf_01(root,0);
    }
}

File: 1022-sum-of-root-to-leaf-binary-numbers/1022-sum-of-root-to-leaf-binary-numbers.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int sumRootToLeaf_01(TreeNode* root , int sum) {
        if(root == nullptr) return 0;
        
        sum = (2*sum) + root->val;
        if(root->left == nullptr && root->right == nullptr) {
            return sum;
        }
        
        return sumRootToLeaf_01(root->left,sum) + sumRootToLeaf_01(root->right,sum);
    }
    int sumRootToLeaf(TreeNode* root) {
        
        return sumRootToLeaf_01(root,0);
    }
};